File: 121.Best-Time-to-Buy-and-Sell-Stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
#
# algorithms
# Easy (46.23%)
# Total Accepted:    439,733
# Total Submissions: 951,179
# beats 100.0% of python submissions


class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        length = len(prices)

        if length < 2:
            return 0

        cur_min, res = prices[0], 0

        for i in xrange(1, length):
            res = max(res, prices[i] - cur_min)
            cur_min = min(cur_min, prices[i])

        return res

File: 121.Best-Time-to-Buy-and-Sell-Stock.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
//
// algorithms
// Easy (47.14%)
// Total Accepted: 491,819
// Total Submissions: 1,043,382


class Solution {
    public int maxProfit(int[] prices) {
        int res = 0;
        int cur_min = Integer.MAX_VALUE;

        for (int p : prices) {
            res = Math.max(p - cur_min, res);
            cur_min = Math.min(cur_min, p);
        }

        return res;
    }
}